### PR TITLE
53978 Weekly NM Download All

### PR DESCRIPTION
**Pull request by mohamme15315**
Coding
Unit test
Integration test
Automation Test

### CODE REVIEW

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 1]
```diff
@@ -382,27 +382,27 @@ public async Task<byte[]> DownloadFssFileAsync(string batchId, string fileName,
 
         }
 
-        public async Task<byte[]> DownloadFSSZipFileAsync(string batchId, string fileName, string correlationId)
+        public async Task<byte[]> DownloadFSSZipFileAsync(string batchId, string fileName, bool isDistributorUser, string correlationId)
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:385]: No need to pass this isDistributorUser parameter from controller as distributor token can be used for downloading both partner & public zip files. Please revert this change.
```suggestion
        public async Task<byte[]> DownloadFSSZipFileAsync(string batchId, string fileName, string correlationId)
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 2]
```diff
@@ -382,27 +382,27 @@ public async Task<byte[]> DownloadFssFileAsync(string batchId, string fileName,
...
         {
             try
             {
-                _logger.LogInformation(EventIds.GetDailyZipNMFileStarted.ToEventId(), "Maritime safety information request to get daily zip NM file started for batchId:{batchId} and fileName:{fileName} with _X-Correlation-ID:{correlationId}", batchId, fileName, correlationId);
+                _logger.LogInformation(EventIds.GetNMZipFileStarted.ToEventId(), "Maritime safety information request to get NM zip file started for batchId:{batchId} and fileName:{fileName} with _X-Correlation-ID:{correlationId}", batchId, fileName, correlationId);
 
-                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(_userService.IsDistributorUser, correlationId);
+                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(isDistributorUser, correlationId);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:391]: As above, please revert this change as well.
```suggestion
                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(_userService.IsDistributorUser, correlationId);
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs [hunk 3]
```diff
@@ -10,7 +10,8 @@ public interface INMDataService
         Task<ShowWeeklyFilesResponseModel> GetWeeklyFilesResponseModelsAsync(int year, int week, string correlationId);
         Task<byte[]> DownloadFssFileAsync(string batchId, string fileName, string correlationId, string frequency);
         Task<ShowNMFilesResponseModel> GetLeisureFilesAsync(string correlationId);
-        Task<byte[]> DownloadFSSZipFileAsync(string batchId, string fileName, string correlationId);
+        Task<byte[]> DownloadFSSZipFileAsync(string batchId, string fileName, bool isDistributorUser, string correlationId);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs:13]: No need to pass isDistributorUser parameter from input.
```suggestion
        Task<byte[]> DownloadFSSZipFileAsync(string batchId, string fileName, string correlationId);
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs [hunk 4]
```diff
@@ -10,7 +10,8 @@ public interface INMDataService
...
+        //Task<byte[]> DownloadFSSAllZipFileAsync(string batchId, string fileName, bool isDistributorUser, string correlationId);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs:14]: Please remove commented line of code.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/Index.cshtml [hunk 5]
```diff
@@ -69,6 +69,7 @@
 
 <script type="text/javascript">
     var serverData = @Html.Raw(Newtonsoft.Json.JsonConvert.SerializeObject(@Model.YearAndWeekList));
+    console.log(@Html.Raw(Newtonsoft.Json.JsonConvert.SerializeObject(@Model.YearAndWeekList)));
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/Index.cshtml:72]: Please remove this console.log statement. I believe it was used for local testing purpose.

**Review by usha13529 — CHANGES_REQUESTED** [review]
Few review comments to fix please.

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Controllers/NoticestoMarinersControllerTest.cs [hunk 6]
```diff
@@ -471,6 +479,53 @@ public async Task WhenAnnualIsCalledAndExceptionThrownByService_ThenShouldReturn
...
+            const string type = "public";
+
+            A.CallTo(() => _fakeNMDataService.DownloadFSSZipFileAsync(A<string>.Ignored, A<string>.Ignored, A<bool>.Ignored, A<string>.Ignored));
+            IActionResult result = await _controller.DownloadAllWeeklyFile(batchId, fileName, mimeType, type);
+
+            Assert.IsInstanceOf<FileResult>(result);
+        }
+
+        [Test]
+        public void WhenDownloadAllWeeklyFileIsCalled_ThenShouldReturnShowDailyFilesAction()
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Controllers/NoticestoMarinersControllerTest.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Logging/EventIds.cs [hunk 7]
```diff
@@ -696,7 +696,19 @@ public enum EventIds
         /// <summary>
         /// 910168 - Maritime safety information request to get banner notification message from azure table failed.
         /// </summary>
-        GetBannerNotificationMessageFromTableFailed = 910168
+        GetBannerNotificationMessageFromTableFailed = 910168,
+        /// <summary>
+        /// 910169 -Request For Download Daily NM File Started.
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Logging/EventIds.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Logging/EventIds.cs [hunk 8]
```diff
@@ -696,7 +696,19 @@ public enum EventIds
...
+        /// </summary>
+        DownloadAllWeeklyNMFileStarted = 910169,
+        /// <summary>
+        /// 910170 -Request For Download Daily NM File Completed.
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Logging/EventIds.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Logging/EventIds.cs [hunk 9]
```diff
@@ -696,7 +696,19 @@ public enum EventIds
...
+        /// </summary>
+        DownloadAllWeeklyNMFileCompleted = 910170,
+        /// <summary>
+        /// 910171 -Request For Download Daily NM File Failed.
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Logging/EventIds.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.IntegrationTests/NoticesToMariners/NoticesToMarinersControllersTest.cs [hunk 10]
```diff
@@ -335,5 +335,32 @@ public async Task WhenCallAnnualWithDuplicateData_ThenReturnUniqueAnnualFiles()
             Assert.AreEqual("---", responseModel.ShowFilesResponseModel[14].Hash);
             Assert.AreEqual("1", responseModel.ShowFilesResponseModel[1].Hash);
         }
+
+        [Test]
+        public async Task WhenCallDownloadAllWeeklyFile_ThenReturnFile()
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.IntegrationTests/NoticesToMariners/NoticesToMarinersControllersTest.cs]: Renamed to WhenCallDownloadAllWeeklyZipFile_ThenReturnZipFile

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 1 above)
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 2 above)
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs (hunk 3 above)
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs (hunk 4 above)
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/Interfaces/INMDataService.cs]: Removed commented line

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/Index.cshtml (hunk 5 above)
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/Index.cshtml]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs [hunk 11]
```diff
@@ -286,5 +289,37 @@ public async Task<FileResult> DownloadDailyFile(string batchId, string fileName,
             }
         }
 
+        public async Task<FileResult> DownloadAllWeeklyFile(string batchId, string fileName, string mimeType, string type)
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs [hunk 12]
```diff
@@ -286,5 +289,37 @@ public async Task<FileResult> DownloadDailyFile(string batchId, string fileName,
...
+            {
+                NMHelper.ValidateParametersForDownloadSingleFile(new()
+                {
+                    new KeyValuePair<string, string>("BatchId", batchId),
+                    new KeyValuePair<string, string>("FileName", fileName),
+                    new KeyValuePair<string, string>("MimeType", mimeType),
+                    new KeyValuePair<string, string>("Type", type)
+                }, GetCurrentCorrelationId(), _logger);
+
+                _logger.LogInformation(EventIds.DownloadAllWeeklyNMFileStarted.ToEventId(), "Maritime safety information request to all weekly NM files started for {type} with batchId:{batchId} and fileName:{fileName} for _X-Correlation-ID:{correlationId}", type, batchId, fileName, GetCurrentCorrelationId());
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs [hunk 13]
```diff
@@ -286,5 +289,37 @@ public async Task<FileResult> DownloadDailyFile(string batchId, string fileName,
...
+
+                byte[] fileBytes;
+                bool isDistributorUser = type == "partner";
+
+                fileBytes = await _nMDataService.DownloadFSSZipFileAsync(batchId, fileName, isDistributorUser, GetCurrentCorrelationId());
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/ShowWeeklyFilesList.cshtml [hunk 14]
```diff
@@ -12,7 +12,33 @@
...
+                        }
+                    }
+                }
+                else
+                {
+                    @if (Model != null && Model.Count > 0)
+                    {
+                        <a id="downloadAll" href="@Url.Action("DownloadAllWeeklyFile", "NoticesToMariners", new { fileName = @ViewData["Year"] + " Wk " + @ViewData["Week"] +" Weekly NMs.zip", batchId = @Model[0].BatchId, mimeType = "application/gzip", type = "public" })">Download All</a>
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (mohamme15315) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/ShowWeeklyFilesList.cshtml]: Done

**Review by usha13529 — CHANGES_REQUESTED** [review]
Just one minor review comment to address please.

**Review by usha13529 — APPROVED** [review]
Thanks @mohamme15315 & @Pradeep15558. Looks good to me.

**Review by prasad07b — APPROVED** [review]
Looks good. Approved.

**Review by Chownder — APPROVED** [review]
Looks good to me, @mohamme15315 @jigar15426 & @Pradeep15558
